# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. 

In [2]:
austin = pd.read_csv("austin_weather.csv")
austin.columns = austin.columns.str.lower().str.replace(" ","_")
df = austin.copy()

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [3]:
df.head(5)

date  temphighf  tempavgf  templowf dewpointhighf dewpointavgf  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  dewpointlowf humidityhighpercent humidityavgpercent humiditylowpercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  sealevelpressureavginches sealevelpressurelowinches visibilityhighmiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  visibilityavgmiles visibilitylowmiles windhighmph windavgmph windgustmph  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  precipitationsuminches               events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

In [4]:
df.shape

(1319, 21)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   date                        1319 non-null   object
 1   temphighf                   1319 non-null   int64 
 2   tempavgf                    1319 non-null   int64 
 3   templowf                    1319 non-null   int64 
 4   dewpointhighf               1319 non-null   object
 5   dewpointavgf                1319 non-null   object
 6   dewpointlowf                1319 non-null   object
 7   humidityhighpercent         1319 non-null   object
 8   humidityavgpercent          1319 non-null   object
 9   humiditylowpercent          1319 non-null   object
 10  sealevelpressurehighinches  1319 non-null   object
 11  sealevelpressureavginches   1319 non-null   object
 12  sealevelpressurelowinches   1319 non-null   object
 13  visibilityhighmiles         1319 non-null   obje

In [6]:
num = df.select_dtypes("number")
cat = df.select_dtypes(exclude="number")

In [7]:
num.shape

(1319, 3)

In [8]:
cat.shape

(1319, 18)

In [9]:
num.describe().round(2)

temphighf  tempavgf  templowf
count    1319.00   1319.00   1319.00
mean       80.86     70.64     59.90
std        14.77     14.05     14.19
min        32.00     29.00     19.00
25%        72.00     62.00     49.00
50%        83.00     73.00     63.00
75%        92.00     83.00     73.00
max       107.00     93.00     81.00

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [10]:
df["dewpointavgf"].unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [11]:
len(df["dewpointavgf"].value_counts())

66

    The dataframe uses "dewpointavgf" as an object because it treats it as a categorical variable 

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [12]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [13]:
wrong_type_columns = [column.lower() for column in wrong_type_columns]

In [14]:
def col_to_numeric(column):
    return pd.to_numeric(column, errors="coerce")

In [15]:
df[wrong_type_columns] = df[wrong_type_columns].apply(col_to_numeric)

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [16]:
df.dtypes

date                           object
temphighf                       int64
tempavgf                        int64
templowf                        int64
dewpointhighf                 float64
dewpointavgf                  float64
dewpointlowf                  float64
humidityhighpercent           float64
humidityavgpercent            float64
humiditylowpercent            float64
sealevelpressurehighinches    float64
sealevelpressureavginches     float64
sealevelpressurelowinches     float64
visibilityhighmiles           float64
visibilityavgmiles            float64
visibilitylowmiles            float64
windhighmph                   float64
windavgmph                    float64
windgustmph                   float64
precipitationsuminches        float64
events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [17]:
missing_values = df.isnull().any(axis=1)

In [18]:
missing_values

0       False
1       False
2       False
3       False
4        True
        ...  
1314    False
1315    False
1316    False
1317    False
1318    False
Length: 1319, dtype: bool

In [19]:
df_w_nulls = df[missing_values]

In [20]:
df_w_nulls

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
4     2013-12-25         58        50        41           44.0          40.0   
6     2013-12-27         60        53        45           41.0          39.0   
7     2013-12-28         62        51        40           43.0          39.0   
42    2014-02-01         76        66        55           62.0          59.0   
51    2014-02-10         60        48        35           49.0          36.0   
...          ...        ...       ...       ...            ...           ...   
1269  2017-06-12         94        85        75           72.0          68.0   
1288  2017-07-01         98        89        79           76.0          72.0   
1304  2017-07-17         98        88        77           75.0          71.0   
1308  2017-07-21        104        91        77           74.0          68.0   
1313  2017-07-26        103        91        78           75.0          70.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
4             36.0                 86.0                71.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
42            41.0                 81.0                71.0   
51            30.0                 82.0                74.0   
...            ...                  ...                 ...   
1269          62.0                 87.0                62.0   
1288          66.0                 85.0                62.0   
1304          66.0                 88.0                63.0   
1308          61.0                 85.0                56.0   
1313          61.0                 91.0                58.0   

      humiditylowpercent  ...  sealevelpressureavginches  \
4                   56.0  ...                      30.33   
6                   47.0  ...                      30.39   
7                   36.0  ...                      30.17   
42                  60.0  ...                      29.81   
51                  66.0  ...                      30.15   
...                  ...  ...                        ...   
1269                36.0  ...                      29.89   
1288                38.0  ...                      30.02   
1304                38.0  ...                      29.90   
1308                27.0  ...                      29.95   
1313                25.0  ...                      30.03   

      sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
4                         30.27                 10.0                10.0   
6                         30.34                 10.0                 9.0   
7                         30.04                 10.0                10.0   
42                        29.75                 10.0                10.0   
51                        30.02                 10.0                 8.0   
...                         ...                  ...                 ...   
1269                      29.81                 10.0                10.0   
1288                      29.92                 10.0                 9.0   
1304                      29.84                 10.0                10.0   
1308                      29.86                 10.0                10.0   
1313                      29.93                 10.0                10.0   

      visibilitylowmiles  windhighmph  windavgmph  windgustmph  \
4                    7.0         10.0         2.0         16.0   
6                    7.0          7.0         1.0         11.0   
7                    7.0         10.0         2.0         14.0   
42                   9.0         14.0         6.0         26.0   
51                   4.0         15.0         9.0         23.0   
...                  ...          ...         ...          ...   
1269                 2.0         15.0         7.0         22.0   
1288                 6.0         15.0         7.0         24.0   
1304                10.0         16.0         3.0         23.0   


There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [21]:
df.shape[0]


1319

In [22]:
df_w_nulls.shape[0]

136

#### Calculate the ratio of missing rows to total rows

In [23]:
ratio_missing_rows = 136/1319*100
ratio_missing_rows


10.310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [24]:
df.isna().sum()

date                            0
temphighf                       0
tempavgf                        0
templowf                        0
dewpointhighf                   7
dewpointavgf                    7
dewpointlowf                    7
humidityhighpercent             2
humidityavgpercent              2
humiditylowpercent              2
sealevelpressurehighinches      3
sealevelpressureavginches       3
sealevelpressurelowinches       3
visibilityhighmiles            12
visibilityavgmiles             12
visibilitylowmiles             12
windhighmph                     2
windavgmph                      2
windgustmph                     4
precipitationsuminches        124
events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [25]:
ratio_nulls_precipi = 124/1319*100
ratio_nulls_precipi

9.401061410159212

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [26]:
df.drop("precipitationsuminches", axis=1, inplace=True)

In [27]:
df.head(1)

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  2013-12-21         74        60        45           67.0          49.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   

   sealevelpressurehighinches  sealevelpressureavginches  \
0                       29.86                      29.68   

   sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                      29.59                 10.0                 7.0   

   visibilitylowmiles  windhighmph  windavgmph  windgustmph  \
0                 2.0         20.0         4.0         31.0   

                events  
0  Rain , Thunderstorm

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [28]:
df_fixed = df.interpolate(inplace=False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [29]:
df_fixed.isnull().sum()

date                          0
temphighf                     0
tempavgf                      0
templowf                      0
dewpointhighf                 0
dewpointavgf                  0
dewpointlowf                  0
humidityhighpercent           0
humidityavgpercent            0
humiditylowpercent            0
sealevelpressurehighinches    0
sealevelpressureavginches     0
sealevelpressurelowinches     0
visibilityhighmiles           0
visibilityavgmiles            0
visibilitylowmiles            0
windhighmph                   0
windavgmph                    0
windgustmph                   0
events                        0
dtype: int64

In [30]:
df.isnull().sum()

date                           0
temphighf                      0
tempavgf                       0
templowf                       0
dewpointhighf                  7
dewpointavgf                   7
dewpointlowf                   7
humidityhighpercent            2
humidityavgpercent             2
humiditylowpercent             2
sealevelpressurehighinches     3
sealevelpressureavginches      3
sealevelpressurelowinches      3
visibilityhighmiles           12
visibilityavgmiles            12
visibilitylowmiles            12
windhighmph                    2
windavgmph                     2
windgustmph                    4
events                         0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [31]:
df.events.value_counts()

events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

[Fog , Rain , Thunderstorm]


#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [32]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

In [33]:
df_fixed["snow"] = 0
df_fixed["fog"] = 0
df_fixed["rain"] = 0
df_fixed["thunderstorm"] = 0

In [34]:
df_fixed.head(1)

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  2013-12-21         74        60        45           67.0          49.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   

   ...  visibilityavgmiles  visibilitylowmiles  windhighmph  windavgmph  \
0  ...                 7.0                 2.0         20.0         4.0   

   windgustmph               events  snow  fog  rain thunderstorm  
0         31.0  Rain , Thunderstorm     0    0     0            0  

[1 rows x 24 columns]

In [35]:
for index,row in df_fixed.iterrows():
    if "Rain" in row["events"]:
        df_fixed.loc[index, "rain"] += 1
    if "Fog" in row["events"]:
        df_fixed.loc[index, "fog"] += 1
    if "Snow" in row["events"]:
        df_fixed.loc[index, "rain"] += 1
    if "Thunderstorm" in row["events"]:
        df_fixed.loc[index, "thunderstorm"] += 1

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [36]:
# I did it by iterrows


#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [37]:
df_fixed.sample(10)

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
969   2016-08-16         83        77        71           76.0          73.0   
505   2015-05-10         87        82        76           72.0          71.0   
294   2014-10-11         78        69        59           72.0          62.0   
6     2013-12-27         60        53        45           41.0          39.0   
510   2015-05-15         77        72        67           71.0          68.0   
1060  2016-11-15         86        73        59           58.0          55.0   
795   2016-02-24         65        54        42           38.0          30.0   
462   2015-03-28         84        66        48           49.0          44.0   
1150  2017-02-13         78        71        63           64.0          55.0   
153   2014-05-23         84        78        71           71.0          67.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
969           70.0                100.0                90.0   
505           69.0                 87.0                71.0   
294           55.0                 97.0                85.0   
6             37.0                 83.0                65.0   
510           65.0                100.0                87.0   
1060          51.0                 87.0                60.0   
795           23.0                 76.0                49.0   
462           40.0                 86.0                54.0   
1150          42.0                 84.0                63.0   
153           62.0                 84.0                69.0   

      humiditylowpercent  ...  visibilityavgmiles  visibilitylowmiles  \
969                 79.0  ...                 6.0                 0.0   
505                 55.0  ...                 9.0                 7.0   
294                 73.0  ...                 7.0                 0.0   
6                   47.0  ...                 9.0                 7.0   
510                 74.0  ...                 8.0                 1.0   
1060                32.0  ...                10.0                10.0   
795                 21.0  ...                10.0                10.0   
462                 22.0  ...                10.0                10.0   
1150                42.0  ...                10.0                10.0   
153                 54.0  ...                 8.0                 0.0   

      windhighmph  windavgmph  windgustmph                     events  snow  \
969          10.0         3.0         16.0  Fog , Rain , Thunderstorm     0   
505          17.0        10.0         29.0                                0   
294          14.0         5.0         20.0                 Fog , Rain     0   
6             7.0         1.0         11.0                                0   
510          13.0         4.0         20.0        Rain , Thunderstorm     0   
1060          8.0         2.0         14.0                                0   
795          18.0         7.0         28.0                                0   
462          13.0         5.0         24.0                                0   
1150         18.0         9.0         28.0                                0   
153          14.0         6.0         20.0                 Fog , Rain     0   

      fog  rain thunderstorm  
969     1     1            1  
505     0     0            0  
294     1     1            0  
6       0     0            0  
510     0     1            1  
1060    0     0            0  
795     0     0            0  
462     0     0            0  
1150    0     0            0  
153     1     1            0  

[10 rows x 24 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [38]:
df_fixed.drop("events", axis=1, inplace=True)

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [39]:
df_fixed["date"] = pd.to_datetime(df_fixed["date"])

In [40]:
df_fixed["date"] = df_fixed["date"].apply(lambda date: date.toordinal())

In [41]:
df_fixed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 23 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   date                        1319 non-null   int64  
 1   temphighf                   1319 non-null   int64  
 2   tempavgf                    1319 non-null   int64  
 3   templowf                    1319 non-null   int64  
 4   dewpointhighf               1319 non-null   float64
 5   dewpointavgf                1319 non-null   float64
 6   dewpointlowf                1319 non-null   float64
 7   humidityhighpercent         1319 non-null   float64
 8   humidityavgpercent          1319 non-null   float64
 9   humiditylowpercent          1319 non-null   float64
 10  sealevelpressurehighinches  1319 non-null   float64
 11  sealevelpressureavginches   1319 non-null   float64
 12  sealevelpressurelowinches   1319 non-null   float64
 13  visibilityhighmiles         1319 

#### Print `austin_fixed` to check your `Date` column.

In [42]:
df_fixed.head(5)

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0  ...                 10.0                 7.0                 2.0   
1  ...                 10.0                10.0                 5.0   
2  ...                 10.0                10.0                10.0   
3  ...                 10.0                10.0                 7.0   
4  ...                 10.0                10.0                 7.0   

   windhighmph  windavgmph  windgustmph  snow  fog  rain  thunderstorm  
0         20.0         4.0         31.0     0    0     1             1  
1         16.0         6.0         25.0     0    0     0             0  
2          8.0         3.0         12.0     0    0     0             0  
3         12.0         4.0         20.0     0    0     0             0  
4         10.0         2.0         16.0     0    0     0             0  

[5 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [43]:
X = df_fixed.drop("tempavgf", axis=1)
y = df_fixed["tempavgf"]

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [50]:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [51]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [52]:
print(f'100% of our data: {len(df_fixed)}.')
print(f'70% for training data: {len(X_train)}.')
print(f'30% for test data: {len(X_test)}.')

100% of our data: 1319.
70% for training data: 1055.
30% for test data: 264.


#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [47]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [48]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [49]:
# Your code here:
